### PR TITLE
South Carolina Senior Exemption

### LINKED ISSUE #2120: South Carolina Senior Exemption
SC Senior Exemption = $15,000 - retirement income deduction - military retirement income deduction

### PR DESCRIPTION
**Pull request by hua7450**
Fixes #2120

Co-authored-by: Pinyan <pxu12@users.noreply.github.com>

Thanks for contributing! Please remove any top-level sections that do not apply to your changes.

- [ ] `make format && make documentation` has been run.

# New variable

- [ ] Label field added
- [ ] Documentation field added
- [ ] Unit field added
- [ ] Default value field added if relevant
- [ ] Variable name follows conventions
- [ ] Unit test(s) added
- [ ] Integration test(s) added if relevant
- [ ] Issues this PR fixes linked

## What's changed

Description of the changes here.

# Bug fix

- [ ] Regression test added
- [ ] Regression test passing

## What this fixes and how it's fixed

Description of how this fix works goes here. Link any issues this PR fixes.

### CODE REVIEW

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+description: South Carolina provides an up to  $15,000 exemption for filers above age 65 (and additional $15,000 if spouse above 65)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml:1]:
```suggestion
description: South Carolina allows filers above this age to be eligible for the senior exemption.
```

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+metadata:
+  type: single_amount
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml:3]:
```suggestion
  unit: year
```

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  label: SC senior exemption age threshold
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml:4]:
```suggestion
  label: South Carolina senior exemption age threshold
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml:4]: and other

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/amount.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+description: South Carolina exemption amount per qualifying person (individual and possibly spouse)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/amount.yaml:1]:
```suggestion
description: South Carolina provides the following senior exemption amount for self.
```

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/amount.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/amount.yaml:1]: Add a second spouse_amount.yaml parameter file to represent the spouse amount allowed under the senior exemption -> incorporate in variable

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]
Let me know if you have questions

File: policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_senior_exemption.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        )
+        sc_military_retirement_income_deduction = tax_unit(
+            "sc_military_retirement_income_deduction", period
+        )
+        sc_military_retirement_income_deduction_spouse = tax_unit(
+            "sc_military_retirement_income_deduction_spouse", period
+        )
+
+        # Calculate senior exemption. The exemption can not be less than 0. Add head and spouse together.
+        return max(
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_senior_exemption.py:48]:
```suggestion
        return max_(
```

**Review by PavelMakarchuk — COMMENTED** [review]
.

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml:1]: need 2021 values

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/spouse_amount.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/spouse_amount.yaml:1]: need 2021 values for all parameters

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/spouse_amount.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+description: South Carolina provides the following senior exemption amount for spouse.
+metadata:
+  unit: currency-USD
+  label: South Carolina senior exemption amount
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/spouse_amount.yaml:4]:
```suggestion
  label: South Carolina senior exemption spouse amount
```

**Review by PavelMakarchuk — COMMENTED** [review]
minor change

File: policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+description: South Carolina allows filers above this age to be eligible for the senior exemption.
+metadata:
+  unit: year
+  type: single_amount
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/sc/tax/income/exemptions/senior/age_threshold.yaml:4]:
```suggestion
```

File: policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_senior_exemption.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            "sc_retirement_income_deduction_spouse", period
+        )
+        sc_military_retirement_income_deduction = tax_unit(
+            "sc_military_retirement_income_deduction", period
+        )
+        sc_military_retirement_income_deduction_spouse = tax_unit(
+            "sc_military_retirement_income_deduction_spouse", period
+        )
+
+        # Calculate senior exemption. The exemption can not be less than 0. Add head and spouse together.
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_senior_exemption.py:47]: I wasn't sure about this until looking at the law, so added a comment and split it up for readability.
```suggestion
        # Calculate senior exemption. The exemption can not be less than 0. Add head and spouse together.
        # Per the legal code, this applies separately:
        # "(2) In the case of married taxpayers who file a joint federal income tax return, the reduction required by item (1) applies to each individual separately"
        head_exemption = max_(
            head_eligible * p.amount - head_deductions, 0
        )
        spouse_exemption = max_(
            spouse_eligible * p.spouse_amount - spouse_deductions, 0
        )
        return head_exemption + spouse_exemption
```

File: policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_military_retirement_income_deduction.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+from policyengine_us.model_api import *
+
+
+class sc_military_retirement_income_deduction(Variable):
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_military_retirement_income_deduction.py:4]: Same for retirement income
```suggestion
class sc_military_retirement_income_deduction_head(Variable):
```

File: policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_military_retirement_income_deduction.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+    value_type = float
+    entity = TaxUnit
+    label = "South Carolina military retirement income deduction for self"
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_military_retirement_income_deduction.py:7]:
```suggestion
    label = "South Carolina military retirement income deduction for head"
```

File: policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_senior_exemption.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+        # Get the spouse age, if applicable.
+        age_spouse = tax_unit("age_spouse", period)
+
+        # Determine whether spouse is eligible (>= age 65).
+        joint = filing_status == filing_status.possible_values.JOINT
+        spouse_eligible = ((age_spouse >= p.age_threshold) * joint).astype(int)
+
+        # Get SC retirement income deduction and military retirement income deduction
+        retirement_income_deduction = tax_unit(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/sc/tax/income/exemptions/sc_senior_exemption.py:34]: same for spouse
```suggestion
        head_deductions = add(tax_unit, period, ["sc_retirement_income_deduction_head", "sc_military_retirement_income_deduction_head"])
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**